src/lib/y2storage/proposal/devices_planner_strategies/base.rb

```diff
@@ -0,0 +1,116 @@
+
+module Y2Storage
+  module Proposal
+    module DevicesPlannerStrategies
+      # Abstract base class with common functionalty for different devices
+      # planner strategies.
+      class Base
+        include Yast::Logger
+
+        attr_accessor :settings
+
+        def initialize(settings, devicegraph)
+          @settings = settings
+          @devicegraph = devicegraph
+        end
+
+        # @!method planned_devices(target)
+        #   List of devices (read: partitions or volumes) that need to be
+        #   created to satisfy the settings.
+        #
+        #   @note This method must be reimplemented by derived classes.
+        #
+        #   @see DevicesPlanner#planned_devices
+        #
+        #   @param target [Symbol] :desired, :min
+        #   @return [Array<Planned::Device>]
+        abstract_method :planned_devices
+
+      protected
+
+        # @return [Devicegraph]
+        attr_reader :devicegraph
+
+        # @return [Symbol] :desired or :min
+        attr_reader :target
+
+        # Planned devices needed by the bootloader
+        #
+        # @return [Array<Planned::Device>]
+        def planned_boot_devices(planned_devices)
+          checker = BootRequirementsChecker.new(
+            devicegraph, planned_devices: planned_devices, boot_disk_name: settings.root_device
+          )
+          checker.needed_partitions(target)
+        rescue BootRequirementsChecker::Error => error
+          # As documented, {BootRequirementsChecker#needed_partition} raises this
+          # exception if it's impossible to get a bootable system, even adding
+          # more partitions.
+          raise NotBootableError, error.message
+        end
+
+        # Swap partition that can be reused.
+        #
+        # It returns the smaller partition that is big enough for our purposes.
+        #
+        # @return [Partition]
+        def reusable_swap(required_size)
+          return nil if settings.use_lvm || settings.use_encryption
+
+          partitions = devicegraph.disk_devices.map(&:swap_partitions).flatten
+          partitions.select! { |part| part.size >= required_size }
+          # Use #name in case of #size tie to provide stable sorting
+          partitions.sort_by { |part| [part.size, part.name] }.first
+        end
+
+        # Delete shadowed subvolumes from each planned device
+        def remove_shadowed_subvolumes(planned_devices)
+          planned_devices.each do |device|
+            next unless device.respond_to?(:subvolumes)
+
+            device.shadowed_subvolumes(planned_devices).each do |subvolume|
+              log.info "Subvolume #{subvolume} would be shadowed. Removing it."
+              device.subvolumes.delete(subvolume)
+            end
+          end
+        end
+
+        # Return the total amount of RAM as DiskSize
+        #
+        # @return [DiskSize] current RAM size
+        def ram_size
+          # FIXME: use the .proc.meminfo agent and its MemTotal field
```

Comment:
Is this something to be solved in this PR?

Reply:
Not, it is out of scope.